client/my-sites/plans-features-main/hooks/use-action-callback.ts

```diff
@@ -0,0 +1,184 @@
+					quantity: selectedStorageAddOn.quantity,
+					volume: 1,
+					extra: { feature_slug: selectedStorageAddOn?.featureSlugs?.[ 0 ] },
+				};
+
+			if ( cartItemForPlan ) {
+				processCartItems?.( [
+					cartItemForPlan,
+					...( storageAddOnCartItem ? [ storageAddOnCartItem ] : [] ),
+				] );
+				return;
+			}
+
+			/* 2. Process free plan for checkout */
+			processCartItems?.( null );
+			return;
+		},
+		[ processCartItems ]
+	);
+}
+
+function useGenerateActionCallback( {
+	currentPlan,
+	eligibleForFreeHostingTrial,
+	cartHandler,
+	flowName,
+	intent,
+	showModalAndExit,
+	sitePlanSlug,
+	siteSlug,
+	withDiscount,
+}: {
+	currentPlan: Plans.SitePlan | undefined;
+	eligibleForFreeHostingTrial: boolean;
+	cartHandler?: ( cartItems?: MinimalRequestCartProduct[] | null ) => void;
+	flowName?: string | null;
+	intent?: PlansIntent | null;
+	showModalAndExit?: ( planSlug: PlanSlug ) => boolean;
+	sitePlanSlug?: PlanSlug | null;
+	siteSlug?: string | null;
+	withDiscount?: string;
+} ): UseActionCallback {
+	const freeTrialPlanSlugs = useFreeTrialPlanSlugs( {
+		intent: intent ?? 'default',
+		eligibleForFreeHostingTrial,
+	} );
+	const currentPlanManageHref = useCurrentPlanManageHref();
+	const handleUpgrade = useUpgradeHandler( { siteSlug, withDiscount, cartHandler } );
+
+	return ( { planSlug, cartItemForPlan, selectedStorageAddOn }: UseActionCallbackParams ) => {
+		return () => {
+			const planConstantObj = applyTestFiltersToPlansList( planSlug, undefined );
+			const freeTrialPlanSlug = freeTrialPlanSlugs?.[ planConstantObj.type ];
+
+			const earlyReturn = showModalAndExit?.( planSlug );
+			if ( earlyReturn ) {
+				return;
+			}
+
+			/* 1. Send user to VIP if it's an enterprise plan */
+			if ( isWpcomEnterpriseGridPlan( planSlug ) ) {
+				recordTracksEvent( 'calypso_plan_step_enterprise_click', { flow: flowName } );
+				const vipLandingPageURL = 'https://wpvip.com/wordpress-vip-agile-content-platform';
+				window.open(
+					`${ vipLandingPageURL }/?utm_source=WordPresscom&utm_medium=automattic_referral&utm_campaign=calypso_signup`,
+					'_blank'
+				);
+				return;
+			}
+
+			/* 2. Send user to either manage add-ons or plan in case of current plan selection */
+			if ( sitePlanSlug && currentPlan?.productSlug === planSlug && intent !== 'plans-p2' ) {
+				if ( isFreePlan( planSlug ) ) {
+					page.redirect( `/add-ons/${ siteSlug }` );
+				} else {
+					page.redirect( currentPlanManageHref );
+				}
+				return;
+			}
+
+			/* 3. Handle plan upgrade and plan upgrade tracks events */
+			switch ( ! isFreePlan( planSlug ) ) {
+				case true:
+					recordTracksEvent?.( 'calypso_plan_features_upgrade_click', {
+						current_plan: sitePlanSlug,
+						upgrading_to: planSlug,
+						saw_free_trial_offer: !! freeTrialPlanSlug,
+					} );
+				case false:
+					recordTracksEvent( 'calypso_signup_free_plan_click' );
+				default:
+					handleUpgrade( {
+						cartItemForPlan,
+						selectedStorageAddOn,
+					} );
+			}
```

Comment:
This will probably not work. I think it's a little trickier to achieve, so let's just go back to normal if/else :-)